storage/index/types.go

```diff
@@ -183,42 +179,306 @@ type BlockTickResult struct {
+// batch entry.
+type ForEachWriteBatchEntryFn func(
+	idx int,
+	entry WriteBatchEntry,
+	doc doc.Document,
+	result WriteBatchEntryResult,
+)
+
+// ForEach allows a caller to perform an operation for each batch entry.
+func (b *WriteBatch) ForEach(fn ForEachWriteBatchEntryFn) {
+	for idx, entry := range b.entries {
+		fn(idx, entry, b.docs[idx], entry.Result())
+	}
+}
+
+// ForEachWriteBatchByBlockStartFn allows a caller to perform an operation with
+// reference to a restricted set of the write batch for each unique block
+// start.
+type ForEachWriteBatchByBlockStartFn func(
+	blockStart time.Time,
+	batch *WriteBatch,
+)
+
+// ForEachUnmarkedBatchByBlockStart allows a caller to perform an operation
+// with reference to a restricted set of the write batch for each unique block
+// start for entries that have not been marked completed yet.
+func (b *WriteBatch) ForEachUnmarkedBatchByBlockStart(
+	fn ForEachWriteBatchByBlockStartFn,
+) {
+	// Ensure sorted correctly first
+	b.SortByUnmarkedAndIndexBlockStart()
+
+	// What we do is a little funky but least alloc intensive, essentially we mutate
+	// this batch and then restore the pointers to the original docs after.
+	allEntries := b.entries
+	allDocs := b.docs
+	defer func() {
+		b.entries = allEntries
+		b.docs = allDocs
+	}()
 
-// ForEachBlockStart iterates over the provided WriteBatchEntryByBlockStart, and calls `fn` on each
-// group of elements with the same blockStart.
-func (w WriteBatchEntryByBlockStart) ForEachBlockStart(fn ForEachBlockStartFn) {
 	var (
-		startIdx  = 0
-		lastNanos xtime.UnixNano
+		blockSize      = b.opts.IndexBlockSize
+		startIdx       = 0
+		lastBlockStart xtime.UnixNano
 	)
-	for i := 0; i < len(w); i++ {
-		elem := w[i]
-		if elem.BlockStart != lastNanos {
-			lastNanos = elem.BlockStart
-			// We only want to call the the ForEachBlockStartFn once we have calculated the entire group,
+	for i := range allEntries {
+		if allEntries[i].OnIndexSeries == nil {
+			// Hit a marked done entry
+			b.entries = allEntries[startIdx:i]
+			b.docs = allDocs[startIdx:i]
+			if len(b.entries) != 0 {
+				fn(lastBlockStart.ToTime(), b)
+			}
+			return
+		}
+
+		blockStart := allEntries[i].indexBlockStart(blockSize)
+		if !blockStart.Equal(lastBlockStart) {
+			prevLastBlockStart := lastBlockStart.ToTime()
+			lastBlockStart = blockStart
+			// We only want to call the the ForEachUnmarkedBatchByBlockStart once we have calculated the entire group,
 			// i.e. once we have gone past the last element for a given blockStart, but the first element
 			// in the slice is a special case because we are always starting a new group at that point.
 			if i == 0 {
 				continue
 			}
-			fn(w[startIdx].BlockStart, w[startIdx:i])
+			b.entries = allEntries[startIdx:i]
+			b.docs = allDocs[startIdx:i]
+			fn(prevLastBlockStart, b)
 			startIdx = i
 		}
 	}
+
 	// spill over
-	if startIdx < len(w) {
-		fn(w[startIdx].BlockStart, w[startIdx:])
+	if startIdx < len(allEntries) {
```

Comment:
hm can you do this spill over un-conditionally? can't there be marked success entries in the back?

Reply:
So you only exit the loop if you haven't hit a "done" element yet, which means all the remaining entries haven't been marked for error or success yet (thanks to the sort order).  I can add a comment to this effect perhaps?

Reply:
yep sounds good